### PR TITLE
Add training ability to CLI

### PR DESCRIPTION
**Pull request by jeromelecoq**
The goal of this PR is to add the ability to invoke a training job through the CLI. This is essential to facilitate bash scripts and easy deployment.

### CODE REVIEW

File: deepinterpolation/cli/training.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        # We pass on folder
+        self.args["training_params"]["output_dir"] = \
+            self.args['training_params']['output_path']
+
+        # We convert to old schema
+        self.args['training_params']['nb_gpus'] = 2 * \
+            int(self.args['training_params']['multi_gpus'])
+
+        # Forward parameters to the training agent
+        self.args['training_params']['steps_per_epoch'] = \
```
Comment (danielsf) [inline deepinterpolation/cli/training.py:42]: Changing `self.args` in the code makes me uneasy. Ideally, `self.args` would just be a record of the arguments passed in by the user. That will especially be helpful if you end up exposing `training_params.steps_per_epoch` to the user in the future. You don't want to silently overwrite your work.

Can you expose `training_params.steps_per_epoch` now, default it to `None`, and, at least, enclose this code in an `if self.args['training_params']['steps_per_epoch'] is None` block (the same goes for all of these lines overwriting or modifying `self.args`?

File: deepinterpolation/cli/training.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            json.dump(self.args['network_params'], f, indent=2)
+        self.logger.info(f"wrote {network_json_path}")
+
+        test_generator_json_path = outdir / f"{uid}_test_generator.json"
+        with open(test_generator_json_path, "w") as f:
+            json.dump(self.args['test_generator_params'], f, indent=2)
+        self.logger.info(f"wrote {test_generator_json_path}")
+
+        generator_obj = ClassLoader(generator_json_path)
+        data_generator = generator_obj.find_and_build()(generator_json_path)
```
Comment (danielsf) [inline deepinterpolation/cli/training.py:69]: I gather that this method reads its parameters from that JSON file. I feel like it would be cleaner if the method accepted a dict of args. You could then move the JSON-writing code into `generator_obj.find_and_build()`. In that case, it's not that you're writing the args you want to use to a file and reading them back in, but you passing args to `generator_obj.find_and_build()` and `find_and_build` is writing out its configuration to a log file.

That feels to me like an arrangement that is less likely to read in the wrong parameters by accident (if, for instance, someone ever changes how this JSON file gets written and forgets to change how `generator_obj.find_and_build()` gets called.

File: deepinterpolation/cli/training.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+        test_generator_obj = ClassLoader(test_generator_json_path)
+        data_test_generator = test_generator_obj.find_and_build()(
+            test_generator_json_path
```
Comment (danielsf) [inline deepinterpolation/cli/training.py:73]: Same comment as above about passing in args as a dict (or even explicitly defining the call signature, and having the `find_and_build` argument record its parameters to a log file.

File: deepinterpolation/cli/training.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        network_obj = ClassLoader(network_json_path)
+        data_network = network_obj.find_and_build()(network_json_path)
+
+        training_obj = ClassLoader(training_json_path)
+
+        training_class = training_obj.find_and_build()(
+            data_generator,
+            data_test_generator,
+            data_network,
+            training_json_path
```
Comment (danielsf) [inline deepinterpolation/cli/training.py:85]: Same comment above about dicts versus json paths

File: tests/test_trainor_collection.py [hunk 5]
```diff
@@ -128,4 +128,4 @@ def test_ephys_training(tmp_path):
 
     # Validation is a bit random due to initilization. We check that you get
```
Comment (danielsf) [inline tests/test_trainor_collection.py:129]: Silly question: but there's no way to pass a random number generator around so that you can control the seed that is fed to the optimizer (I assume there is an optimizer somewhere that is the source of the randomness)?

File: deepinterpolation/cli/schemas.py [hunk 6]
```diff
@@ -163,3 +163,123 @@ def inference_specific_settings(self, data, **kwargs):
...
+        description=("type and name sent to ClassLoader for object "
+                     "instantiation"))
+
+    name = argschema.fields.String(
+        required=False,
+        default="core_trainer",
+        description=("type and name sent to ClassLoader for object "
+                     "instantiation"))
+
+    output_path = argschema.fields.String(
```
[the remark this answers is not in the document]
Comment (jeromelecoq) [reply deepinterpolation/cli/schemas.py:181]: I would prefer to keep output_path to keep it consistent with the current parameters terms. I did modify to use the right argshema

File: deepinterpolation/cli/schemas.py [hunk 7]
```diff
@@ -163,3 +163,123 @@ def inference_specific_settings(self, data, **kwargs):
...
+
+    learning_rate = argschema.fields.Float(
+        required=False,
+        default=0.0001,
+        description="base learning rate used by the optimizer")
+
+    loss = argschema.fields.String(
+        required=False,
+        default="mean_squared_error",
+        description="loss function used for training and validation.")
```
[the remark this answers is not in the document]
Comment (jeromelecoq) [reply deepinterpolation/cli/schemas.py:201]: I added more details in the description. Essentially Loss functions recognized by tensorflow are recognized : https://www.tensorflow.org/api_docs/python/tf/keras/losses. Additional losses can be added to the loss_collection.py file. So it is an extendable list.

File: deepinterpolation/cli/schemas.py [hunk 8]
```diff
@@ -163,3 +163,123 @@ def inference_specific_settings(self, data, **kwargs):
...
+
+    output_path = argschema.fields.String(
+        required=True,
+        description="A folder where the training outputs will get written.")
+
+    nb_times_through_data = argschema.fields.Int(
+        required=False,
+        default=1,
+        description="Setting this to more than 1 will make the training use \
+            individual samples multiple time during training, thereby \
```
[the remark this answers is not in the document]
Comment (jeromelecoq) [reply deepinterpolation/cli/schemas.py:189]: Fixed

File: deepinterpolation/cli/training.py (hunk 1 above)
Comment (jeromelecoq) [reply deepinterpolation/cli/training.py:42]: So this parameters is exposed actually. The reason it is passed here is that the intermediate generator and training object needs to have access to this parameter in the current implementation. I am hoping to remove that dependency with a separate PR (as discussed above). Exposing this parameters more than once will force the use to feed the parameter multiple times which cause some errors. I therefore decided to copy it around there.

File: deepinterpolation/cli/training.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            self.args['training_params']["model_string"] = (
+                self.args["network_params"]["name"] + "_" +
+                self.args["training_params"]["loss"]
+            )
+
+        # We pass on the uid
+        self.args["training_params"]["run_uid"] = uid
+
+        # We pass on folder
+        self.args["training_params"]["output_dir"] = \
```
[the remark this answers is not in the document]
Comment (jeromelecoq) [reply deepinterpolation/cli/training.py:34]: I removed this renaming

File: tests/test_trainor_collection.py (hunk 5 above)
Comment (jeromelecoq) [reply tests/test_trainor_collection.py:129]: That is an interesting point. It might go beyond the goal of this PR.  Will not address this but we can look into this with a later PR.

File: deepinterpolation/cli/schemas.py [hunk 10]
```diff
@@ -163,3 +163,123 @@ def inference_specific_settings(self, data, **kwargs):
...
+class NetworkSchema(argschema.schemas.DefaultSchema):
+    type = argschema.fields.String(
+        required=False,
+        default="network",
+        description=("type of object instantiation"))
+    name = argschema.fields.String(
+        required=True,
+        default="unet_single_1024",
+        description=("callback of the neuronal network architecture to build.\
+            All available architectures are visible in the \
```
[the remark this answers is not in the document]
Comment (jeromelecoq) [reply deepinterpolation/cli/schemas.py:341]: I fixed this using the OneOf Suggestion.

File: deepinterpolation/cli/training.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+class Training(argschema.ArgSchemaParser):
+    default_schema = TrainingInputSchema
+
+    def run(self):
+        self.logger.name = type(self).__name__
+
+        outdir = Path(self.args['training_params']['output_path'])
+        if self.args["output_full_args"]:
+            full_args_path = outdir / "training_full_args.json"
```
[the remark this answers is not in the document]
Comment (jeromelecoq) [reply deepinterpolation/cli/training.py:17]: Double checked : All files are appended the uid already by default to protect against this.

File: deepinterpolation/cli/schemas.py [hunk 12]
```diff
@@ -163,3 +163,123 @@ def inference_specific_settings(self, data, **kwargs):
...
+        default=5,
+        description="Period in number of epochs to periodically save model \
+            checkpoints.")
+
+
+class NetworkSchema(argschema.schemas.DefaultSchema):
+    type = argschema.fields.String(
+        required=False,
+        default="network",
+        description=("type of object instantiation"))
```
[the remark this answers is not in the document]
Comment (jeromelecoq) [reply deepinterpolation/cli/schemas.py:336]: Fixed

**Comment by jeromelecoq** [comment]
Added a number of input checking.
